compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -851,40 +902,387 @@ installDepsAndInitPackageDb opts (InitPkgDb shouldInit) =
+  -- Must drop DAML_PROJECT from env var so it can be repopulated based on `cwd`
+  assistantEnv <- filter ((/="DAML_PROJECT") . fst) <$> getEnvironment
+
+  buildableDataDepsMapping <- fmap Map.fromList $ for (mpPackagePaths multiPackageConfig) $ \path -> do
+    darPath <- darPathFromDamlYaml path
+    pure (darPath, path)
+
+  let assistantRunner = AssistantRunner $ \location args ->
+        withCreateProcess ((proc assistantPath args) {cwd = Just location, env = Just assistantEnv})
+          $ \_ _ _ p -> void $ waitForProcess p
+      buildableDataDeps = BuildableDataDeps $ flip Map.lookup buildableDataDepsMapping
+      mRootPkgBuilder = (\pkgConfig -> buildEffect relativize pkgConfig opts mbOutFile incrementalBuild initPkgDb) <$> mPkgConfig
+      mRootPkgData = (toNormalizedFilePath' $ maybe cDir unwrapProjectPath $ projectRoot projectOpts,) <$> mRootPkgBuilder
+      rule = buildMultiRule assistantRunner buildableDataDeps noCache mRootPkgData
+ 
+  -- Set up a near empty shake environment, with just the buildMulti rule
+  bracket
+    (IDE.initialise rule (DummyLspEnv diagnosticsLogger) (toIdeLogger loggerH) noopDebouncer (toCompileOpts opts) vfs)
+    IDE.shutdown
+    $ \ideState -> runActionSync ideState
+      $ case mRootPkgData of
+          Nothing -> void $ uses_ BuildMulti $ toNormalizedFilePath' <$> mpPackagePaths multiPackageConfig
+          Just (rootPkgPath, _) -> void $ use_ BuildMulti rootPkgPath
+
+data AssistantRunner = AssistantRunner { runAssistant :: FilePath -> [String] -> IO ()}
+
+-- Stores a mapping from dar path to project path
+data BuildableDataDeps = BuildableDataDeps { getDataDepSource :: FilePath -> Maybe FilePath }
+
+data BuildMulti = BuildMulti
+    deriving (Eq, Show, Typeable, Generic)
+instance Binary BuildMulti
+instance Hashable BuildMulti
+instance NFData BuildMulti
+
+-- We return all the information needed to derive the dalf name
+type instance RuleResult BuildMulti = (LF.PackageName, LF.PackageVersion, LF.PackageId)
+
+-- Subset of PackageConfig needed for multi-package build deferring.
+data BuildMultiPackageConfig = BuildMultiPackageConfig
+  { bmSdkVersion :: PackageSdkVersion
+  , bmName :: LF.PackageName
+  , bmVersion :: LF.PackageVersion
+  , bmDataDeps :: [FilePath]
+      -- ^ not canonicalized
+  , bmSourceDaml :: FilePath
+      -- ^ not canonicalized
+  , bmOutput :: Maybe FilePath
+      -- ^ not canonicalized
+  }
+
+-- Version of getDamlFiles from Dar.hs that assumes source is a folder, and runs in IO rather than action
+-- Throws an error if its a path to a file
+getDamlFilesBuildMulti :: FilePath -> IO (Map.Map FilePath BSL.ByteString)
+getDamlFilesBuildMulti path = do
+  exists <- doesDirectoryExist path
+  -- TODO[SW]: find a way to fix this
+  -- It isn't fully possible, since we need to be able to parse the files to check their imports, and we can't parse other sdk versions.
+  -- The best we can probably do is find all daml files in the daml.yaml directory?
```

Comment:
We now simply return nothing from the function (so the file staleness check always fails) and warn the user about this.
Fortunately, since the package-id won't change, the staleness won't propagate up the dependency tree